### PR TITLE
Update Ruff

### PR DESCRIPTION
**Pull request by alexauritt**
This PR updates ruff and also updates the code base to comply with some new checks, most importantly the check that asks to move imports inside a `if TYPE_DEFINITION` block.

Related: #2477

### CODE REVIEW

File: strawberry/aiohttp/handlers/graphql_ws_handler.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
 from contextlib import suppress
-from typing import Any, Optional
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Any, Optional
```
Comment (patrick91) [inline strawberry/aiohttp/handlers/graphql_ws_handler.py:5]: do you if there's a way to skip having these imports behind `if TYPE_CHECKING`?

feels a bit annoying to have to do that all the time for things inside typing (we still need to import TYPE_CHECKING anyway) :D

**Comment by patrick91** [comment]
By the failure on the tests:

```
  File "/home/runner/work/strawberry/strawberry/strawberry/utils/operation.py", line 14, in <module>
    graphql_document: DocumentNode,
NameError: name 'DocumentNode' is not defined
```

I think we need to add `from __future__ import annotations` at the top of the files that use `if TYPE_CHECKING`

**Comment by patrick91** [comment]
can you also push the updated `pyproject.toml`? :)

File: strawberry/aiohttp/handlers/graphql_ws_handler.py (hunk 1 above)
Comment (alexauritt) [reply strawberry/aiohttp/handlers/graphql_ws_handler.py:5]: agree, quite annoying. I'm not sure if there's a way. I'll pause the work and do some digging. Thanks for flagging.

Comment (alexauritt) [reply strawberry/aiohttp/handlers/graphql_ws_handler.py:7]: https://github.com/charliermarsh/ruff/issues/2267

Comment (alexauritt) [reply strawberry/aiohttp/handlers/graphql_ws_handler.py:7]: looks like this is addressed in latest release of `ruff` - https://github.com/charliermarsh/ruff/releases/tag/v0.0.237

Comment (alexauritt) [reply strawberry/aiohttp/handlers/graphql_ws_handler.py:7]: I'll rollback those changes, thanks again for flagging before I went too deep

**Comment by alexauritt** [comment]
> can you also push the updated `pyproject.toml`? :)

added but rolled it back for now to keep the branch green. I'll re-update the toml once ruff is satisfied

**Comment by alexauritt** [comment]
@patrick91 -  I think this PR is ready to come out of draft mode. I should note that while my original intention was just to enable `ANN001` checks (as part of https://github.com/strawberry-graphql/strawberry/issues/2477), this PR doesn't do that at all! I got confused + distracted by the TCH errors I was seeing, and I only just realized that this was because I was running a later version of ruff locally. 

So... I guess this branch has turned into an 'upgrade ruff' PR...

Is that helpful or welcome? Should I open a new issue?

I'll be happy to follow up with work on the ANN001

Thanks!

**Comment by alexauritt** [comment]
Also - would be happy to keep plugging away and include the ANN001 checks on this branch if you prefer, but thought I should check in given the size of this PR

**Comment by patrick91** [comment]
hi @alexauritt, sorry for the late reply! I'll check this in the next couple of days!
thanks so much for all the work you're putting in this PR 😊

**Comment by alexauritt** [comment]
@patrick91 - I should acknowledge that I'm still seeing [that same error](https://github.com/strawberry-graphql/strawberry/actions/runs/4086912711/jobs/7134627292) surfacing when running tests. A bit stumped by that.

**Comment by patrick91** [comment]
@alexauritt I'll probably review this on saturday during [our sprints](https://beta.guild.host/events/strawberry-graphql-sprints-wzz8u1) :D

**Comment by alexauritt** [comment]
If you've any hints as to how I could help tackle the test failures let me know.

**Comment by alexauritt** [comment]
Hoping to join the sprint, virtually, for a of couple hours at least

**Comment by patrick91** [comment]
Hi @alexauritt, I took the liberty of doing some commits on the PR!

One issue was that the changes were also touching code that was needed at runtime (I forgot what it was now)

Another issue was that we needed `from __future__ import annotations` in a lot of places, since the types are behind the `if TYPE_CHECKING` flag (using future annotations makes python treat annotations as string, not code)

When using `from __future__ import annotations` `casts` that reference an object imported inside the `if TYPE_CHECKING` flag need to use quotes around them like so:

```python
if TYPE_CHECKING:
   from a import A

x = cast("A", 123)
```

on top of this `from __future__ import annotations` does break some of our code (in the federation/schema.py file). I haven't investigated why just yet 😊

And fastapi dependencies need to be always imported (so they can't be in the `if TYPE_CHECKING` flag)

I think that was all of it 😊

File: strawberry/ext/mypy_plugin.py [hunk 2]
```diff
@@ -337,7 +354,7 @@ def add_static_method_to_class(
             cls.defs.body.remove(sym.node)
 
     # For compat with mypy < 0.93
-    if MypyVersion.VERSION < Decimal("0.93"):
+    if Decimal("0.93") > MypyVersion.VERSION:
```
Comment (patrick91) [inline strawberry/ext/mypy_plugin.py:357]: I'll revert this change, I don't think I like it :D

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Alex Auritt for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Review by patrick91 — APPROVED** [review]
Thanks @alexauritt for doing this! I think this should good to merge!

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**